src/test/java/uk/gov/companieshouse/acsp/manage/users/integration/UserAcspMembershipIntegrationTest.java

```diff
@@ -710,6 +720,882 @@ void addAcspMemberSuccessWhenInviteeHasRemovedMembershipInDifferentAcsp() throws
+                    .contentType( MediaType.APPLICATION_JSON )
+                    .content( "{\"action\":\"edit_role\",\"user_role\":\"standard\"}" ) )
+            .andExpect( status().isBadRequest() );
+  }
+
+  @Test
+  void updateAcspMembershipForIdWhereStandardAttemptsToChangeStandardToStandardReturnsBadRequest() throws Exception {
+    final var acspMemberDaos = testDataManager.fetchAcspMembersDaos( "XME003", "XME004" );
+
+    acspMembersRepository.insert( acspMemberDaos );
+    Mockito.doReturn( testDataManager.fetchUserDtos( "67ZeMsvAEgkBWs7tNKacdrPvOmQ" ).getFirst() ).when( usersService ).fetchUserDetails( "67ZeMsvAEgkBWs7tNKacdrPvOmQ" );
+
+    mockMvc.perform( patch( "/acsp-members/XME003" )
+                    .header("X-Request-Id", "theId123")
+                    .header("Eric-identity", "67ZeMsvAEgkBWs7tNKacdrPvOmQ")
+                    .header("ERIC-Identity-Type", "oauth2")
+                    .header("ERIC-Authorised-Key-Roles", "*")
+                    .contentType( MediaType.APPLICATION_JSON )
+                    .content( "{\"action\":\"edit_role\",\"user_role\":\"standard\"}" ) )
+            .andExpect( status().isBadRequest() );
+  }
+
+  @Test
+  void updateAcspMembershipForIdReturnsBadRequestWhenOwnerAttemptsToRemoveThemselves() throws Exception {
+    acspMembersRepository.insert( testDataManager.fetchAcspMembersDaos( "WIT004" ) );
+    Mockito.doReturn( testDataManager.fetchUserDtos( "67ZeMsvAEgkBWs7tNKacdrPvOmQ" ).getFirst() ).when( usersService ).fetchUserDetails( "67ZeMsvAEgkBWs7tNKacdrPvOmQ" );
+
+    mockMvc.perform( patch( "/acsp-members/WIT004" )
+                    .header("X-Request-Id", "theId123")
+                    .header("Eric-identity", "67ZeMsvAEgkBWs7tNKacdrPvOmQ")
+                    .header("ERIC-Identity-Type", "oauth2")
+                    .header("ERIC-Authorised-Key-Roles", "*")
+                    .contentType( MediaType.APPLICATION_JSON )
+                    .content( "{\"action\":\"remove_member\"}" ) )
+            .andExpect( status().isBadRequest() );
+  }
+
+  @Test
+  void updateAcspMembershipForIdReturnsBadRequestWhenOwnerAttemptsToChangeTheirOwnRole() throws Exception {
+    acspMembersRepository.insert( testDataManager.fetchAcspMembersDaos(  "WIT004" ) );
+    Mockito.doReturn( testDataManager.fetchUserDtos( "67ZeMsvAEgkBWs7tNKacdrPvOmQ" ).getFirst() ).when( usersService ).fetchUserDetails( "67ZeMsvAEgkBWs7tNKacdrPvOmQ" );
+
+    mockMvc.perform( patch( "/acsp-members/WIT004" )
+                    .header("X-Request-Id", "theId123")
+                    .header("Eric-identity", "67ZeMsvAEgkBWs7tNKacdrPvOmQ")
+                    .header("ERIC-Identity-Type", "oauth2")
+                    .header("ERIC-Authorised-Key-Roles", "*")
+                    .contentType( MediaType.APPLICATION_JSON )
+                    .content( "{\"action\":\"edit_role\", \"user_role\":\"standard\" }" ) )
+            .andExpect( status().isBadRequest() );
+  }
```

Comment:
I originally missed these tests as the file was folded, but re-reviewing now, I think we could parametrize these.  I think the comprehensive test coverage is good, but there's a huge amount of duplicated code here and parameterization would improve the readability a LOT.

I.e: parsing through this is really difficult

<img width="773" alt="Screenshot 2024-07-09 at 11 12 58" src="https://github.com/companieshouse/acsp-manage-users-api/assets/170090135/ff7164e1-7650-4c71-83b9-cecd3aaae9c6">

Reply:
Here's an example (would need to refactor your existing tests):

```java
class AcspMembershipUpdateTests {

    private static Stream<Arguments> membershipUpdateScenarios() {
        return Stream.of(
            arguments("Owner", "Owner", "Admin", true, status().isOk()),
            arguments("Owner", "Admin", "Standard", true, status().isOk()),
            arguments("Owner", "Standard", "Admin", true, status().isOk()),
            arguments("Admin", "Standard", "Admin", true, status().isOk()),
            arguments("Admin", "Admin", "Standard", true, status().isOk()),
            arguments("Admin", "Owner", "Admin", false, status().isBadRequest()),
            arguments("Standard", "Owner", "Admin", false, status().isBadRequest()),
            arguments("Standard", "Admin", "Standard", false, status().isBadRequest()),
            arguments("Standard", "Standard", "Admin", false, status().isBadRequest())
            // Add more scenarios as needed
        );
    }

    @ParameterizedTest
    @MethodSource("membershipUpdateScenarios")
    void testUpdateAcspMembership(String actorRole, String targetRole, String newRole, 
                                  boolean shouldSucceed, ResultMatcher expectedStatus) throws Exception {

        // Setup actors, or use testDataManager.fetchAcspMembersDaos
        String actorId = setupActor(actorRole);
        String targetId = setupTarget(targetRole);

        mockMvc.perform(patch("/acsp-members/" + targetId)
                .header("X-Request-Id", "theId123")
                .header("Eric-identity", actorId)
                .header("ERIC-Identity-Type", "oauth2")
                .header("ERIC-Authorised-Key-Roles", "*")
                .contentType(MediaType.APPLICATION_JSON)
                .content("{\"action\":\"edit_role\",\"user_role\":\"" + newRole.toLowerCase() + "\"}"))
                .andExpect(expectedStatus);

        if (shouldSucceed) {
            AcspMemberDao updatedMember = acspMembersRepository.findById(targetId).orElseThrow();
            assertEquals(UserRoleEnum.valueOf(newRole.toUpperCase()), updatedMember.getUserRole());
        }
    }

    private String setupActor(String role) { ... }
    private String setupTarget(String role) { ... }

    // Similar parameterized tests can be created for remove_member action
}
```

Reply:
I also find a glance at `membershipUpdateScenarios` is really helpful for learning the proper scenarios.